Add the httptrace feature

tkanos opened the pull request:
In order to be able to trace the latency of the connection. I use the httptrace package from golang.
If trace is nil, we can set a default one, that trace the basics (you can add more callback functions if you want, this basic was define by looking what was needed on https://github.com/tcnksm/go-httpstat)

On zipkin the big overview will give you some dots (with hover labels) :

![zipkin span view](https://lh3.googleusercontent.com/4Axn8DI43vvaTblvif40UFT4cU2UHX04PJfyBGapabZIkUUqkLyhZKtMyTujZdJ6qv5QIWPQGH-uGRMYLkEI=w1853-h913)

And if you open the span, you will see each log, and the time when it occured (allowing you to see the latency) :

![zipkin span log](https://lh3.googleusercontent.com/4vJoIxsPhrZdGAARC-WhflFtbLt8f7Q6umHIKJaYwJpM0hHZ1n-4YsXCiUqp7b-HWS5SAR1dWziwtRc5SxY_=w1853-h913)

tkanos commented:
If you want to test, start a zipkin server :

```bash
docker run -d -p 9411:9411 openzipkin/zipkin
```

and then execute the following code (being on the branch  httptrace-integration)
```golang
package main

import (
	"context"
	"net/http"
	"os"
	"time"

	"github.com/ricardo-ch/go-tracing"
)

// this name is use to identify traces inside zipkin UI
const (
	appName = "my-application"
)

func main() {
	tracing.SetGlobalTracer(appName, "http://localhost:9411")
	defer tracing.FlushCollector()

	doWork(context.Background())
}

func doWork(ctx context.Context) {
	span, ctx := tracing.CreateSpan(ctx, "doWork", nil)
	defer span.Finish()

	r, err := googleRequest(ctx)
	if err != nil {
		os.Exit(-1)
	}
	// Add The trace to the Request
	r = tracing.InjectHTTPTraceSpan(r, nil)
	client := &http.Client{
		Timeout: time.Duration(100) * (time.Millisecond),
	}
	client.Do(r)
}

func googleRequest(ctx context.Context) (r *http.Request, er error) {
	r, er = http.NewRequest("GET", "http://www.google.ca/", nil) // ;)
	if er != nil {
		return nil, er

	}
	r = r.WithContext(ctx)

	return r, nil
}
```

FrancoisPoinsot commented on span.go:
```diff
@@ -110,3 +112,18 @@ func InjectSpan(r *http.Request) *http.Request {
 }
+
+// InjectHTTPTraceSpan Injects HttpTrace on the request
+func InjectHTTPTraceSpan(r *http.Request, trace *httptrace.ClientTrace) *http.Request {
```
This seem super useful. So much that i don't see the point not doing it every time we call InjectSpan() .

Why not make this private and insert it in injectSpan

FrancoisPoinsot replied:
also this `httptrace.ClientTrace` would nicely fit with the option pattern.
This would reduce signature to `InjectHTTPTraceSpan(r *http.Request, option Options...) ` and make it obvious that it is totally ok the not give any ClientTrace when calling this method.

FrancoisPoinsot replied:
Combine these two and you won't have to add any line to the existing example to use this feature. 
Which mean still having super simple framework which is the point of this repo.

tkanos replied:
Hello @FrancoisPoinsot,

I like the options pattern idea, but do you prefer the options pattern or the functional options pattern ?

FrancoisPoinsot replied:
Now that i checked i only know what functional option pattern look like. No idea what option pattern is without these function call

FrancoisPoinsot commented:
Sorry for the delay. That is indeed a nice touch. Would be better if we could not make this lib more difficult to use.

tkanos commented:
Hello @FrancoisPoinsot,

It was a super idea, I added the tracer as Optional.
and I also added an example folder.

Let me know if you are OK, with that implementation ?

tkanos commented:
Hello @FrancoisPoinsot,

Are you sure that you want to automatically inject httptrace on InjectSpan ?
The only thing I see against (and it's a small thing), is if the user wants to inject his own httptrace options. He will have to call implicitely injectHTTPTrace twice.
One on the InjectSpan and then explicitly InjectHTTPTraceSpan to override what was done implicitely.

If we do so the code will be 
```golang

//InjectSpan ...
func InjectSpan(r *http.Request) *http.Request {
	// Retrieve the Span from context
	if span := opentracing.SpanFromContext(r.Context()); span != nil {
		// We are going to use this span in a client request, so mark as such.
		ext.SpanKindRPCClient.Set(span)
		// Retrieve tracer
		tracer := opentracing.GlobalTracer()
		// Inject the Span context into the outgoing HTTP Request
		tracer.Inject(
			span.Context(),
			opentracing.TextMap,
			opentracing.HTTPHeadersCarrier(r.Header),
		)
		// Inject HttpTrace by default on the request
		r = injectHTTPTrace(r, span)
	}
	return r
}

// InjectHTTPTraceSpan Injects HttpTrace on the request
func InjectHTTPTraceSpan(r *http.Request, options ...func(*httptrace.ClientTrace, opentracing.Span)) *http.Request {

	if span := opentracing.SpanFromContext(r.Context()); span != nil {
		r = injectHTTPTrace(r, span, options...)
	}

	return r
}

func injectHTTPTrace(r *http.Request, span opentracing.Span, options ...func(*httptrace.ClientTrace, opentracing.Span)) *http.Request {
	trace := newClientTrace(span)
	for _, option := range options {
		option(trace, span)
	}

	ctx := httptrace.WithClientTrace(r.Context(), trace)
	r = r.WithContext(ctx)

	return r
}
```

FrancoisPoinsot commented:
You are right, having 2 HTTPTraceClient working would be problematic

Was more thinking something such as https://github.com/lovoo/goka/blob/master/processor.go#L59
 `options` would be part of all public call, even `InjectSpan` . in theses `options` you would find a `WithTraceClient()`

Code would be more like 

```
func InjectSpan(r *http.Request, options ...Options) *http.Request {
	// Retrieve the Span from context
	if span := opentracing.SpanFromContext(r.Context()); span != nil {
		// We are going to use this span in a client request, so mark as such.
		ext.SpanKindRPCClient.Set(span)
		// Retrieve tracer
		tracer := opentracing.GlobalTracer()
		// Inject the Span context into the outgoing HTTP Request
		tracer.Inject(
			span.Context(),
			opentracing.TextMap,
			opentracing.HTTPHeadersCarrier(r.Header),
		)
		// Inject HttpTrace by default on the request
		traceClient := options.WithClientTrace(newClientTrace(span))
		r = injectHTTPTrace(r, span, traceClient)
	}
	return r
}

func injectHTTPTrace(r *http.Request, span opentracing.Span, traceClient HTTPTraceClient) *http.Request {
	ctx := httptrace.WithClientTrace(r.Context(), traceClient)
	r = r.WithContext(ctx)

	return r
}
```

But then again, maybe it is too much and just having a nullable HttpTraceClient given through call is good enough with a bit of comment explaining that if `nil` default one would be used.

edit: i removed the `InjectHTTPTraceSpan` because i really don't see any reason not to use both trace and context passing by default when observing an httpRequest

tkanos commented:
Hello @FrancoisPoinsot, 

It makes sense to remove InjectHTTPTraceSpan, and use directly InjectSpan to inject the variadic options. I have done the changes.

FrancoisPoinsot commented:
Sorry that it took more than a month to handle this.
This definitely show that we are lacking guidelines on this project. Will be added shortly. (shorter than a month)

In the mean time thanks for this particularly good feature which don't even require to change anything client side. And that is beautiful!
